### PR TITLE
feat: add transaction hash to zero trace

### LINKED ISSUE #97: Fix parsing of the latest Erigon witness format
There is a new tx hash field in the latest [Erigon](https://github.com/0xPolygonZero/erigon) json output from the `debug_traceBlockByNumber`

### PR DESCRIPTION
**Pull request by atanmarko**
With the Erigon [commit](https://github.com/0xPolygonZero/erigon/commit/2002fde2289407a3ee23aeeaa710c292191b2a32) output of `debug_traceBlockByNumber` has changed, so now tx result json looks like:
```
{
	"txHash":"0x058c80b157b4fb4f8b925a81d384336e79afaa64820fbd133c73b774bf6953e7",
	"result":{"traces":{...},"meta": {...}}
}
``` 
Block witness returned from `eth_getWitness` now looks like:
```
"0x00055821033601462093b5945d1676df093446790fd31b20e7b12a2e8e5e09d068109616b0084a021e19e0c9bab240000005582103468288056310c82aa4c01a7e12a10f8111a0560e72b700555479031b86c357d0084101055820021df1fa259221d02aa4956eb0d35ace318ca24c0a33a64c1af96cf67cf245b6084101055820021703c5eda8644a64cec152c58f5aacec93d72fb0bfa705f0473f9043a8357c08410102188405582103b70e80538acdabd6137353b0f9d8d149f4dba91e8be2e7946e409bfdbe685b900841010558210389802d6ed1a28b049e9d4fe5334c5902fd9bc00c42821c82f82ee2da10be90800841010558200256274a27dd7524955417c11ecd917251cc7c4c8310f4c7e4bd3c304d3d9a79084a021e19e0c9bab2400000055820023ab0970b73895b8c9959bae685c3a19f45eb5ad89d42b52a340ec4ac204d190841010219102005582103876da518a393dbd067dc72abfa08d475ed6447fca96d92ec3f9e7eba503ca6100841010558210352688a8f926c816ca1e079067caba944f158e764817b83fc43594370ca9cf62008410105582103690b239ba3aaf993e443ae14aeffc44cf8d9931a79baed9fa141d0e4506e131008410102196573"
```

Deserialization logic in zero-bin has changed accordingly.
Trace decoder uses `primitive-types` H256 to represent hash, but as we are switching to `Alloy` we chose here `alloy::primitives::TxHash` for tx hash type.

Resolves https://github.com/0xPolygonZero/zero-bin/issues/97

Block witness is now retrieved with a separate `eth_getWitness` rpc call.

### CODE REVIEW

File: rpc/src/lib.rs [hunk 1]
```diff
@@ -42,20 +47,18 @@ where
...
 
-    let (txn_info, mut pre_images) =
-        traces
-            .into_iter()
-            .partition_map::<Vec<_>, Vec<_>, _, _, _>(|it| match it {
-                ZeroTrace::Result(it) => Either::Left(it),
-                ZeroTrace::BlockWitness(it) => Either::Right(it),
-            });
+    // Grab block witness info (packed as combined trie pre-images)
+    ///////////////////////////////////////////////////////////////
```
Comment (muursh) [inline rpc/src/lib.rs]: Not convinced by the lines of `////`. I know they already exist in the code base though

**Review by muursh — APPROVED** [review]
Looks fine to me. Is it tested?

File: rpc/src/lib.rs (hunk 1 above)
Comment (atanmarko) [reply rpc/src/lib.rs]: Just followed the existing code. Happy to remove them.

Comment (Nashtare) [reply rpc/src/lib.rs]: Yeah I had told Aatif on his PR that this wasn't the convention we had on `zero-bin` but it got merged nonetheless. I am not against it (not my preference though), but I wouldn't mind too much as long as we get consistent across the whole codebase.

Comment (muursh) [reply rpc/src/lib.rs]: I'm happy for a vote on this to take the consensus on the `zero_dev` channel.

**Comment by atanmarko** [comment]
> Looks fine to me. Is it tested?

@muursh Yes it is tested with the current [Erigon feat/zero](https://github.com/0xPolygonZero/erigon). Merging would also require update of zk_evm dependency to current develop branch, are there any obstacles for that?

Latest `zk_evm` has also some new parameter `code_db` in the BlockTrace prover input. Not sure how to initialize that (for test I just put default value)?

**Comment by Nashtare** [comment]
@atanmarko  `code_db` can be defaulted to `None` for `Jerigon`. This is an add-on for the native tracer incoming through #81

File: Cargo.toml [hunk 2]
```diff
@@ -26,9 +26,9 @@ alloy = { git = "https://github.com/alloy-rs/alloy", features = [
 
 # zk-evm dependencies
 plonky2 = "0.2.2"
-evm_arithmetization = { git = "https://github.com/0xPolygonZero/zk_evm.git", tag = "v0.3.1" }
-trace_decoder = { git = "https://github.com/0xPolygonZero/zk_evm.git", tag = "v0.3.1" }
-proof_gen = { git = "https://github.com/0xPolygonZero/zk_evm.git", tag = "v0.3.1" }
+evm_arithmetization = { git = "https://github.com/0xPolygonZero/zk_evm.git", branch = "develop" }
```
Comment (muursh) [inline Cargo.toml]: It's probably worth waiting until @Nashtare does the release for this

Comment (Nashtare) [reply Cargo.toml]: I already have the new versions ready. I can re-run the entire Ethereum test suite + John's 200 Shanghai L1 blocks that we saved to check for regression and we could do the release if it's pressing.

Comment (muursh) [reply Cargo.toml]: Sounds sensible.

**Review by Nashtare — APPROVED** [review]

File: Cargo.toml (hunk 2 above)
Comment (Nashtare) [reply Cargo.toml]: The Ethereum test suite passed fine on the `zk_evm@prep_release` branch, as well as all John's L1 block witnesses (apart from two that OOMed, retrying them as I write this). So we should be good with rolling the new versions soon.

**Comment by Nashtare** [comment]
@atanmarko New versions have been published :) Feel free to swap to tag `v0.4.0` in your dependencies.

**Pull request merged**